pkg/kubectl/cmd/push.go

```diff
@@ -0,0 +1,212 @@
+		Use:     "push image [IMAGE_NAME]",
+		Short:   "Push image to cluster reigstry.",
+		Long:    "Push image to cluster reigstry.",
+		Example: push_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunPush(f, cmd, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	// Default registry name "kube-registry" and port "5000", which matches addon configs.
+	cmd.Flags().StringP("host", "H", default_endpoint, "Docker host to connect to")
+	cmd.Flags().StringP("registry", "r", default_registry_name, "Registry service name")
+	cmd.Flags().StringP("port", "p", default_registry_port, "Registry service port")
+	return cmd
+}
+
+func RunPush(f *cmdutil.Factory, cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return cmdutil.UsageError(cmd, "IMAGE is required for pushing")
+	}
+	imageName := args[0]
+	endpoint := cmdutil.GetFlagString(cmd, "host")
+	registryName := cmdutil.GetFlagString(cmd, "registry")
+	port := cmdutil.GetFlagString(cmd, "port")
+
+	// Find the image and tag it with localhost.
+	dockerClient, err := docker.NewClient(endpoint)
+	if err != nil {
+		return err
+	}
+
+	host, repo, tag := parseRepositoryTag(imageName)
+	if len(host) != 0 {
+		fmt.Printf("Host %v will be rewritten to localhost:%s\n", host, port)
+	}
+	if len(tag) == 0 {
+		tag = "latest"
+	}
+
+	tagOpts := docker.TagImageOptions{
+		Repo:  fmt.Sprintf("localhost:%s/%s", port, repo),
+		Tag:   tag,
+		Force: true,
+	}
+	err = dockerClient.TagImage(imageName, tagOpts)
+	if err != nil {
+		return err
+	}
+
+	// Find registry service and pods.
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+
+	service, err := client.Services(api.NamespaceSystem).Get(registryName)
+	if err != nil {
+		return err
+	}
+
+	pods, err := client.Pods(api.NamespaceSystem).List(labels.SelectorFromSet(service.Spec.Selector), fields.Everything())
+	if err != nil {
+		return err
+	}
+
+	// Pick a random pod to start forwarding.
+	found := false
+	seed := rand.Intn(len(pods.Items))
+	var pod api.Pod
+	for i := 0; i < len(pods.Items); i++ {
+		pod = pods.Items[(i+seed)%len(pods.Items)]
+		if pod.Status.Phase != api.PodRunning {
+			continue
+		} else {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return cmdutil.UsageError(cmd, "Unable to find a running registry pod")
+	}
+
+	// Start portfoward for between localhost and remote registry pod.
+	config, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	stopCh := make(chan struct{}, 1)
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
+
+	go func() {
+		<-signals
+		close(stopCh)
+	}()
+
+	// Start port-forwarding for the pods.
+	fw := &defaultPortForwarder{}
+	req := client.RESTClient.Post().
+		Resource("pods").
+		Namespace(api.NamespaceSystem).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	go fw.ForwardPorts("POST", req.URL(), config, []string{port}, stopCh)
+	// TODO: Wait until portforwarder starts.
+	time.Sleep(3 * time.Second)
```

Comment:
Have to wait portforwarder before pushing.  I can look into a better approach if the kubectl command is accepted.